consensus/consensus-types/src/block.rs

```diff
@@ -372,7 +369,7 @@ where
         if self.is_genesis_block() {
             return Ok(());
         }
-        debug_checked_verify_eq!(self.id(), self.hash(), "Block id mismatch the hash");
+        ensure!(self.id() == self.hash(), "Block id mismatch the hash");
```

Comment:
So, IIUC, when using lcs through annotations only, we re-introduce the id field in the serialized form. And this is why you're changing the debug into an ensure.

This does not have to be. We can have a manual `Serialize` and `Deserialize` implementation for the `Block` type only, where `Serialize` "forgets" the id field, and `Deserialize` re-adds it safely.

This would me markedly safer for this sensitive data structure.

Reply:
I suppose if the id was sent incorrectly that the sig check would fail.

There's a number of ways to do what @huitseeker mentions in a pretty easy way (without completely writing a Serialize and Deserialize impl from scratch). Give me a little bit and I'll share an example

Reply:
So you can do something like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a7c1fe1fcbd8a1da3ba4ec26f173aee6) to still be able to benefit from the Derives.

Reply:
Thanks @bmwill, I think this is the most efficient way to keep the existing behavior!

Reply:
Thanks :)

Reply:
We may still want to make this a MIRAI annotation (irrespective of debug)